src/main/java/com/varchenko/MikeHashMap.java

```diff
@@ -0,0 +1,282 @@
+
+
+import java.util.*;
+
+// Node -> null, Node -> key null, Node -> key, value
+public class MikeHashMap<K, V> extends AbstractMap<K, V> implements Map<K, V> {
+    private static final float DEFAULT_LOAD_FACTOR = 0.75f;
+    private static final int DEFAULT_CAPACITY = 16;
+    private Node<K, V>[] buckets;
+    private float loadFactor;
+    private int capacity;
+    private int modCount;
+    private int size;
+
+    public MikeHashMap() {
+        this.capacity = DEFAULT_CAPACITY;
+        buckets = (Node<K, V>[]) new Node[capacity];
+        this.loadFactor = DEFAULT_LOAD_FACTOR;
+    }
+
+    public MikeHashMap(int capacity, float loadFactor) {
+        this();
+        this.capacity = capacity;
+        this.loadFactor = loadFactor;
+    }
+
+    private void resize() {
+        if (size >= capacity * loadFactor) {
+            Node<K, V>[] oldBucket = buckets;
+            buckets = new Node[capacity + (capacity >> 1)];
+            for (Node<K, V> kvNode : oldBucket) {
+                if (kvNode != null) {
+                    put(kvNode.getKey(), kvNode.getValue());
+                }
+            }
+        }
+    }
+
+    private int indexFor(Object k) {
+        if (k != null) {
+            int index = k.hashCode() & capacity - 1;
+            return index >= 0 ? index : -index;
+        }
+        return 0;
+    }
+
+    private Node<K, V> getHead(K key) {
+        return buckets[indexFor(key)];
+    }
+
+    private Node<K, V> find(Object key) {
+        return key != null ? findNotNull((K) key) : findNull();
+    }
+
+    private Node<K, V> findNotNull(K key) {
+        Node<K, V> head = getHead(key);
+        if (head != null) {
+            while (head.getKey() != null) {
+                if (head.getKey().hashCode() == key.hashCode() && head.getKey().equals(key)) {
+                    return head;
+                }
+            }
+        }
+        return null;
+    }
+
+    private Node<K, V> findNull() {
+        Node<K, V> head = buckets[0];
+        if (head != null) {
+            while (head.hasNext()) {
+                head = head.next;
+                if (head.getKey() == null) {
+                    return head;
+                }
+            }
+            return head;
+        }
+        return null;
+    }
+
+    @Override
+    public V get(Object key) {
+        Node<K, V> currentNode = find(key);
+        if (currentNode != null) {
+            return currentNode.getValue();
+        }
+        return null;
+    }
+
+    @Override
+    public V put(K key, V value) {
+        resize();
+        Node<K, V> head = getHead(key);
+        if (head == null) {
+            createHead(key, value);
+        } else {
+            return putEntry(head, key, value);
+        }
+        return null;
+    }
+
+    private V putEntry(Node<K, V> head, K key, V value) {
+        Node<K, V> cursor = head;
+        while (cursor.hasNext()) {
+            cursor = cursor.next;
+            if (key != null) {
```

Comment:
Why do you check (each iteration) that the key is not null?

Reply:
done